networks/rpc/http_newrelic.go

```diff
@@ -0,0 +1,136 @@
+package rpc
+
+import (
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"runtime/debug"
+
+	"github.com/gin-gonic/gin/binding"
+	"github.com/newrelic/go-agent/v3/newrelic"
+)
+
+// dupWriter writes data to the buffer as well as http response
+type dupWriter struct {
+	http.ResponseWriter
+	body *bytes.Buffer
+}
+
+func (w dupWriter) Write(b []byte) (int, error) {
+	w.body.Write(b)
+	return w.ResponseWriter.Write(b)
+}
+
+// KASAttrs contains identifications for a KAS request
+type KASAttrs struct {
+	ChainID      string `header:"x-chain-id" json:"x-chain-id"`
+	AccountID    string `header:"x-account-id" json:"x-account-id"`
+	RequestID    string `header:"x-request-id" json:"x-request-id"`
+	ParentSpanID string `header:"x-b3-parentspanid,omitempty" json:"x-b3-parentspanid,omitempty"`
+	SpanID       string `header:"x-b3-spanid,omitempty" json:"x-b3-spanid,omitempty"`
+	TraceID      string `header:"x-b3-traceid,omitempty" json:"x-b3-traceid,omitempty"`
+}
+
+func parseKASHeader(r *http.Request) KASAttrs {
+	attrs := KASAttrs{}
+	if err := binding.Header.Bind(r, &attrs); err != nil {
+		logger.Error("failed to bind a KAS header")
+	}
+	return attrs
+}
+
+// newNewRelicHTTPHandler enables NewRelic web transaction monitor.
+// It also prints error logs when RPC returns contains error messages.
+func newNewRelicHTTPHandler(nrApp *newrelic.Application, handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		defer func() {
+			if err := recover(); err != nil {
+				logger.Error("NewRelic http handler panic\n"+string(debug.Stack()), "err", err)
+			}
+		}()
+
+		// parse RPC requests
+		reqs, isBatch, err := getRPCRequests(r)
+		if err != nil || len(reqs) < 1 {
+			logger.Error("failed to parse RPC request", "err", err, "len(reqs)", len(reqs))
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		// new relic transaction name contains the first API method of the request
+		reqMethod := reqs[0].method
+		if isBatch {
+			reqMethod += "_batch"
+		}
+
+		txn := nrApp.StartTransaction(r.Method + " " + r.URL.String() + " " + reqMethod)
+		defer txn.End()
+
+		w = txn.SetWebResponse(w)
+		txn.SetWebRequestHTTP(r)
+		r = newrelic.RequestWithTransactionContext(r, txn)
+
+		// duplicate writer
+		dupW := &dupWriter{
+			ResponseWriter: w,
+			body:           bytes.NewBufferString(""),
+		}
+
+		// serve HTTP
+		handler.ServeHTTP(dupW, r)
+
+		// print RPC error logs if errors exist
+		if isBatch {
+			var rpcReturns []interface{}
+			if err := json.Unmarshal(dupW.body.Bytes(), &rpcReturns); err == nil {
+				for i, rpcReturn := range rpcReturns {
+					if data, err := json.Marshal(rpcReturn); err == nil {
+						printRPCErrorLog(data, reqs[i].method, r)
+					}
+				}
+			}
+		} else {
+			printRPCErrorLog(dupW.body.Bytes(), reqs[0].method, r)
+		}
```

Comment:
Can't we always unmarshal `dupW.body.Bytes()` with `var rpcReturns []interface{}`? (meaning always take `isBatch` branch`) 
```suggestion
		var rpcReturns []interface{}
		if err := json.Unmarshal(dupW.body.Bytes(), &rpcReturns); err == nil {
			for i, rpcReturn := range rpcReturns {
				if data, err := json.Marshal(rpcReturn); err == nil {
					printRPCErrorLog(data, reqs[i].method, r)
				}
			}
		}
```

Reply:
Thank you for the kind suggestion. However, it will fail to parse a single request. If the request is single, `for i, rpcReturn := range rpcReturns {` will iterate with keys of the json return. For example, `rpcReturn` will be `jsonrpc`, `id`, or `result` with your code and the following return. `json` is unmashalled as `[]interface` internally. and `the array of json` is unmarshalled as `[][]interface`.
```
{
    "jsonrpc": "2.0",
    "id": 73,
    "result": "0x0"
}
```